Technological Institute of the Philippines | Quezon City - Computer Engineering
--- | ---
Course Code: | CPE 019
Code Title: | Emerging Technologies in CpE 2
2nd Semester | AY 2023-2024
<hr> | <hr>
<u>**Assignment 5.2** | Build and Apply Multilayer Perceptron
**Name** | Buenafe, Dhafny S.
**Section** | CPE32S3
**Date Performed**: |March 26, 2024
**Date Submitted**: |March 26, 2024
**Instructor**: | Engr. Roman Richard

<hr>

In this assignment, you are task to build a multilayer perceptron model. The following are the requirements:

- Choose any dataset
- Explain the problem you are trying to solve
- Create your own model
- Evaluate the accuracy of your model

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
df_wine = pd.read_csv('/content/Wine.csv')


The problem in this dataset that we are trying to solve is to predict the quality of Vinho Verde, a type of wine, based on the various physicochemical attributes provided in the dataframe.

In [ ]:
df_wine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  Id  
0      9.4        5   0  
1      9.8        5   1  
2      9.8        5   2  
3      9.8        6   3  
4      9.4        5   4

In [ ]:
df_wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1143 non-null   float64
 1   volatile acidity      1143 non-null   float64
 2   citric acid           1143 non-null   float64
 3   residual sugar        1143 non-null   float64
 4   chlorides             1143 non-null   float64
 5   free sulfur dioxide   1143 non-null   float64
 6   total sulfur dioxide  1143 non-null   float64
 7   density               1143 non-null   float64
 8   pH                    1143 non-null   float64
 9   sulphates             1143 non-null   float64
 10  alcohol               1143 non-null   float64
 11  quality               1143 non-null   int64  
 12  Id                    1143 non-null   int64  
dtypes: float64(11), int64(2)
memory usage: 116.2 KB


In [ ]:
df_wine.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
Id                      0
dtype: int64

In [ ]:
df_wine.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1143.000000       1143.000000  1143.000000     1143.000000   
mean        8.311111          0.531339     0.268364        2.532152   
std         1.747595          0.179633     0.196686        1.355917   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.392500     0.090000        1.900000   
50%         7.900000          0.520000     0.250000        2.200000   
75%         9.100000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1143.000000          1143.000000           1143.000000  1143.000000   
mean      0.086933            15.615486             45.914698     0.996730   
std       0.047267            10.250486             32.782130     0.001925   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             21.000000     0.995570   
50%       0.079000            13.000000             37.000000     0.996680   
75%       0.090000            21.000000             61.000000     0.997845   
max       0.611000            68.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality           Id  
count  1143.000000  1143.000000  1143.000000  1143.000000  1143.000000  
mean      3.311015     0.657708    10.442111     5.657043   804.969379  
std       0.156664     0.170399     1.082196     0.805824   463.997116  
min       2.740000     0.330000     8.400000     3.000000     0.000000  
25%       3.205000     0.550000     9.500000     5.000000   411.000000  
50%       3.310000     0.620000    10.200000     6.000000   794.000000  
75%       3.400000     0.730000    11.100000     6.000000  1209.500000  
max       4.010000     2.000000    14.900000     8.000000  1597.000000

In [ ]:
#Independent Variable
y = pd.get_dummies(df_wine["quality"])
y

3  4  5  6  7  8
0     0  0  1  0  0  0
1     0  0  1  0  0  0
2     0  0  1  0  0  0
3     0  0  0  1  0  0
4     0  0  1  0  0  0
...  .. .. .. .. .. ..
1138  0  0  0  1  0  0
1139  0  0  0  1  0  0
1140  0  0  1  0  0  0
1141  0  0  0  1  0  0
1142  0  0  1  0  0  0

[1143 rows x 6 columns]

In [ ]:
#Dependent Variable
x = df_wine.drop(["quality", "Id"], axis=1)
x

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1138            6.3             0.510         0.13             2.3      0.076   
1139            6.8             0.620         0.08             1.9      0.068   
1140            6.2             0.600         0.08             2.0      0.090   
1141            5.9             0.550         0.10             2.2      0.062   
1142            5.9             0.645         0.12             2.0      0.075   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1138                 29.0                  40.0  0.99574  3.42       0.75   
1139                 28.0                  38.0  0.99651  3.42       0.82   
1140                 32.0                  44.0  0.99490  3.45       0.58   
1141                 39.0                  51.0  0.99512  3.52       0.76   
1142                 32.0                  44.0  0.99547  3.57       0.71   

      alcohol  
0         9.4  
1         9.8  
2         9.8  
3         9.8  
4         9.4  
...       ...  
1138     11.0  
1139      9.5  
1140     10.5  
1141     11.2  
1142     10.2  

[1143 rows x 11 columns]

**SPLITTING DATA**

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.4)

In [ ]:
x_train

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
975             8.7              0.78         0.51             1.7      0.415   
1121            6.2              0.51         0.14             1.9      0.056   
543             8.8              0.42         0.21             2.5      0.092   
569             7.1              0.61         0.02             2.5      0.081   
200             7.7              0.27         0.68             3.5      0.358   
...             ...               ...          ...             ...        ...   
37              8.6              0.38         0.36             3.0      0.081   
978             7.5              0.57         0.02             2.6      0.077   
855             7.2              0.36         0.46             2.1      0.074   
493             5.1              0.47         0.02             1.3      0.034   
1122            6.4              0.36         0.53             2.2      0.230   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
975                  12.0                  66.0  0.99623  3.00       1.17   
1121                 15.0                  34.0  0.99396  3.48       0.57   
543                  33.0                  88.0  0.99823  3.19       0.52   
569                  17.0                  87.0  0.99745  3.48       0.60   
200                   5.0                  10.0  0.99720  3.25       1.08   
...                   ...                   ...      ...   ...        ...   
37                   30.0                 119.0  0.99700  3.20       0.56   
978                  11.0                  35.0  0.99557  3.36       0.62   
855                  24.0                  44.0  0.99534  3.40       0.85   
493                  18.0                  44.0  0.99210  3.90       0.62   
1122                 19.0                  35.0  0.99340  3.37       0.93   

      alcohol  
975       9.2  
1121     11.5  
543       9.2  
569       9.7  
200       9.9  
...       ...  
37        9.4  
978      10.8  
855      11.0  
493      12.8  
1122     12.4  

[685 rows x 11 columns]

In [ ]:
y_train

3  4  5  6  7  8
975   0  0  1  0  0  0
1121  0  0  0  1  0  0
543   0  0  1  0  0  0
569   0  0  0  1  0  0
200   0  0  0  0  1  0
...  .. .. .. .. .. ..
37    0  0  1  0  0  0
978   0  0  0  1  0  0
855   0  0  0  0  1  0
493   0  0  0  1  0  0
1122  0  0  0  1  0  0

[685 rows x 6 columns]

**MULTILAYER PERCEPTRON**

In [ ]:
from keras.models import Sequential
from keras.layers import Dense

In [ ]:
model = Sequential()

#For number of hidden layers
model.add(Dense(5, activation='sigmoid'))
#For number of  output nodes
model.add(Dense(6, activation='softmax'))

In [ ]:
model.compile(loss='categorical_crossentropy',metrics=["accuracy"])

In [ ]:
#Number of iteration
train_history = model.fit(x_train, y_train, batch_size=25, epochs=50, validation_data=(x_test,y_test))

Epoch 1/50
28/28 [==============================] - 1s 11ms/step - loss: 1.1333 - accuracy: 0.4920 - val_loss: 1.1321 - val_accuracy: 0.5153
Epoch 2/50
28/28 [==============================] - 0s 5ms/step - loss: 1.1316 - accuracy: 0.4920 - val_loss: 1.1318 - val_accuracy: 0.5131
Epoch 3/50
28/28 [==============================] - 0s 3ms/step - loss: 1.1313 - accuracy: 0.4949 - val_loss: 1.1315 - val_accuracy: 0.5240
Epoch 4/50
28/28 [==============================] - 0s 5ms/step - loss: 1.1308 - accuracy: 0.5007 - val_loss: 1.1310 - val_accuracy: 0.5197
Epoch 5/50
28/28 [==============================] - 0s 6ms/step - loss: 1.1303 - accuracy: 0.4964 - val_loss: 1.1307 - val_accuracy: 0.5218
Epoch 6/50
28/28 [==============================] - 0s 5ms/step - loss: 1.1294 - accuracy: 0.4949 - val_loss: 1.1307 - val_accuracy: 0.5197
Epoch 7/50
28/28 [==============================] - 0s 5ms/step - loss: 1.1293 - accuracy: 0.4920 - val_loss: 1.1299 - val_accuracy: 0.5197
Epoch 8/50
28/28 [=

In [ ]:
score = model.evaluate(x_test, y_test, verbose=False)

print('Test score: ', score[0])    #Loss on test
print('Test accuracy: ', score[1])

Test score:  1.1167978048324585
Test accuracy:  0.5218340754508972


#Model Evaluation

The loss test, quantifying the model's overall performance, stands at 1.1167978048324585. The higher the loss test, the poorer the model's performance. The model's accuracy is only 0.5218340754508972, indicating that it's not performing at a high level. This low accuracy can be attributed to the inadequacy of the network structure, as it forms the foundation for achieving higher accuracy. This accuracy level implies that the model correctly predicts the outcome for only about half of the test instances, suggesting room for improvement in creating a model.